### ISSUE TITLE
Physics pause in WebVR presentation mode

### ISSUE THREAD

**Issue by donalffons** [react-spring/use-cannon#99]
Hi,
I have been testing the [react-xr ping pong](https://codesandbox.io/s/react-xr-paddle-demo-v4uet) demo with my Oculus Quest headset. Unfortunately, the physics simulation is not working as soon as I enter the WebVR presentation mode in the default Oculus browser. It does however work in Firefox.

I have done a bit of research into the topic and realized that the the "loop" function is never invoked after the call to [requestAnimationFrame](https://github.com/react-spring/use-cannon/blob/master/src/Provider.tsx#L134), when I'm in presentation mode.

I have done a bit of research and found [this thread](https://github.com/immersive-web/webxr/issues/225), where it is mentioned that the behavior of the Oculus browser (pausing the animation loop) might be the correct behavior: 

> I think we want to conclude with something like "requestAnimationFrame / requestPostAnimationFrame / requestIdleCallback do not fire when the 2d browser view is not visible.

**Comment by codynova** [comment]
This seems related to https://github.com/react-spring/use-cannon/discussions/73, but it sounds like you're saying `use-cannon` runs fine outside of "presentation mode" (while still using the Oculus browser)? I don't have an Oculus and am not familiar with its software.

**Comment by donalffons** [comment]
Yes, right, that seems to be the same issue. And yes, the simulation runs fine when not in WebVR presentation mode. As soon as I enter the WebVR presentation mode, the physics simulation pauses. When I exit the WebVR presentation mode, physics continue normally.

However, I don't think that this problem has anything to do with the fact, that use-cannon runs a web worker. I managed to resolve this issue for my testing purposes by replacing the [call to requestAnimationFrame](https://github.com/react-spring/use-cannon/blob/master/src/Provider.tsx#L134) with the following line
```javascript
setTimeout(loop, 1000/60).
```
Using this code, the physics simulation continues regardless of whether I'm in WebVR presentation mode or not. Works with Oculus browser and in Firefox.

However, I'm not sure if this is the _correct_ approach to solving this issue (feels like its not :-)). I also don't know how to do this the right way... There seems to be a [specific requestAnimationFrame function for WebXR](https://developer.mozilla.org/en-US/docs/Web/API/XRSession/requestAnimationFrame) contexts (which contains a warning in the red box regarding this issue) - maybe this would be the right thing to use? I wouldn't know how to access the `XRSession` object from use-cannon`s Provider, though....

**Comment by codynova** [comment]
Great research. From what I can see on MDN, it doesn't seem like the WebXR API is available from a web worker.

 I don't have a VR device to be able to test, but I think the first step would be checking whether you have access to the `navigator.xr` object from inside the worker. If that works, then I think it would be something like this:

```
const session = navigator.xr.getSession()
session.requestAnimationFrame()
```

* [Navigator API](https://developer.mozilla.org/en-US/docs/Web/API/Navigator)
* [WebXR Device API](https://developer.mozilla.org/en-US/docs/Web/API/WebXR_Device_API)
* [**Functions and classes available to workers**](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Functions_and_classes_available_to_workers)
* [**WorkerNavigator Interface**](https://developer.mozilla.org/en-US/docs/Web/API/WorkerNavigator) - this is the subset of the `navigator` object that's available to workers, and it doesn't contain a reference to the WebXR interface.

Maybe there is another solution. Unfortunately I am not familiar with developing VR on the web.

**Comment by donalffons** [comment]
Hmm, the `navigator.xr` object does not have a method `getSession()`. It does have a method called `requestSession`, which returns a promise... That `requestSession` method seems to be the one that in turn makes the browser ask the user if he wants to start an immersive vr session (hence the promise). When I call `requestSession` there, I get the error message
```
DOMException: Failed to execute 'requestSession' on 'XRSystem': The requested session requires user activation.
```

Coming from a ThreeJS perspective, there is the `xr` object under the [`WebGLRenderer`](https://threejs.org/docs/#api/en/renderers/webxr/WebXRManager). This is of type [WebXRManager and has a `getSession()`](https://github.com/mrdoob/three.js/blob/master/src/renderers/webxr/WebXRManager.js#L163) method.

**Comment by donalffons** [comment]
...unfortunately, simply doing
```javascript
if(gl.xr.isPresenting) {
  gl.xr.getSession().requestAnimationFrame(loop);
} else {
  requestAnimationFrame(loop);
}
```
doesn't work. For some reason, the "physics" loop stops (i.e. never resumes), when changing between VR and non-VR mode. Therefore, I use [this code](https://github.com/donalffons/use-cannon/blob/master/src/Provider.tsx#L115) to restart the physics loop when the presentation mode changes. Firefox and Oculus browser behave differently. Oculus works fine, while Firefox complains about `attempting to access detached arraybuffer`. I assume the `worker.postMessage` line is executed twice in Firefox, which is why I added [these checks](https://github.com/donalffons/use-cannon/blob/master/src/Provider.tsx#L109).

I know, that's not pretty, but it works... Feel free to use this code (or don't :-)). It might be a good idea though, if someone with a different (non-Oculus) headset could test this as well.

**Comment by hitchcott** [comment]
Thanks for the fix @donalffons, I tested your branch and it seems to work.

**Comment by codynova** [comment]
Released #100 in `use-cannon@0.5.1`
